File: 200. Number of Islands/200.cpp
// BFS
// worst case every element in grid is land -> traversing entire graph
// time: O(m*n) 
// space: O(min(m,n)) maximum length of queue: worst case expanding from middle until hit the sorter end then stop appending to queue
class Solution {
public:
    int dirs[4][2] = {{-1, 0}, {0,1}, {1, 0}, {0, -1}};

    int numIslands(vector<vector<char>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        int totnum = 0;
        vector<vector<bool>> vis(m, vector<bool>(n,false));

        for(int i = 0; i < m; i++){
            for (int j = 0; j< n; j++){
                if (grid[i][j] == '1' && !vis[i][j]){
                    totnum += 1;

                    queue<pair<int, int>> bfs_queue;
                    bfs_queue.push({i, j});

                    while (!bfs_queue.empty()){
                        pair<int, int> curnode = bfs_queue.front();
                        int currow = curnode.first;
                        int curcol = curnode.second;
                        vis[currow][curcol] = true; // mark as visited

                        // traverse through all 4 directions
                        for(int i = 0; i< 4; i++){
                            int newrow = currow + dirs[i][0];
                            int newcol = curcol + dirs[i][1];
                            if (newrow >=0 && newrow <m && newcol>=0 && newcol<n){
                                if (grid[newrow][newcol] == '1' && !vis[newrow][newcol]){
                                    // avoid visiting repeating neighboring nodes
                                    vis[newrow][newcol] = true;
                                    bfs_queue.push({newrow, newcol});
                                }
                            }
                        }
                        bfs_queue.pop();
                    }
                }
            }
        }

        return totnum;
    }
};

File: 200. Number of Islands/200.py
"""
depth first search
time: O(m*n)
space: O(m*n) -> recursive stack
"""
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        num_islands = 0
        m = len(grid)
        n = len(grid[0])

        def dfs(cur_row, cur_col):
            if grid[cur_row][cur_col] != '1':
                return

            grid[cur_row][cur_col] = '#'

            row_ch = [0, 1, 0, -1]
            col_ch = [1, 0, -1, 0]

            for i in range(len(row_ch)):
                newrow = cur_row + row_ch[i]
                newcol = cur_col + col_ch[i]

                if newrow >=0 and newrow < m and newcol >=0 and newcol <n:
                    dfs(newrow, newcol)

        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    dfs(i, j)
                    num_islands += 1

        return num_islands

"""
num_islands = 0
def dfs(cur_row, cur_col):
    if grid[cur_row][cur_col] is water:
        return

    grid[cur_row][cur_col] = '#'

    visit each neighbor

traverse through each row, col of grid:
    if current cell is land:
        dfs(row, col)
        num_islands += 1
"""
